### PR TITLE
Fix the IndexOutOfBoundsException when setChildren/manageChildren an illegal child ReactRawTextShadowNode to a parent non-ReactBaseTextShadowNode.

### PR DESCRIPTION
**Pull request by raphael-liu**
## Summary 

Hi, @cpojer , would you please help me reviewing the pull request? Thank you.

* Bad Case:
```
    <View>
       123
       '456'
       <Text>{"setChildren/manageChildren"}</Text>
    </View>
```

* Root Cause:
image

* Solution:
Drop the illegal children node(s) and correct the indices at which other children views to add when setChildren/manageChildren.

* Issues: #18773 #17118 #17113 #16411 #1211 #924

## Changelog:

[Android] [Fixed] -  Fix the IndexOutOfBoundsException when setChildren/manageChildren an illegal child ReactRawTextShadowNode to a parent non-ReactBaseTextShadowNode.

## Test Plan:
----------
See UIImplementationTest.java ( Tests for UIImplementation ).

### CODE REVIEW

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java [hunk 1]
```diff
@@ -44,6 +47,14 @@
  */
 public class UIImplementation {
 
+  static {
+    try {
+      mReactBaseTextShadowNodeClass = Class.forName("com.facebook.react.views.text.ReactBaseTextShadowNode");
```
Comment (raphael-liu) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java]: Without using reflection, to break circular dependency of buck, there will be up to 795 lines changed. And Warnings:
image

**Comment by anchengjian** [comment]
same problem!

File: ReactAndroid/src/test/java/com/facebook/react/uimanager/UIImplementationTest.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
+package com.facebook.react.uimanager;
```
Comment (hramos) [inline ReactAndroid/src/test/java/com/facebook/react/uimanager/UIImplementationTest.java]: Please add here the standard copyright header used elsewhere in the codebase.

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java [hunk 3]
```diff
@@ -357,6 +368,142 @@ protected void handleUpdateView(
...
+    for (int i = 0; i < vais.size(); i++) {
+      ViewAtIndex viewAtIndex = vais.get(i);
+      ReactShadowNode cssNodeToAdd = mShadowNodeRegistry.getNode(viewAtIndex.mTag);
+      if (cssNodeToAdd == null) {
+        throw new IllegalViewOperationException("Trying to add unknown view tag: "
+          + viewAtIndex.mTag);
+      }
+      // If there is a RCTRawText child node, and the parent is not a ReactBaseTextShadowNode, donot add the illegal
+      // RCTRawText childNode to parent node.
+      // Fix bug 1216685-IndexOutOfBoundsException
```
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java]: You can remove the "fix bug" line here, as it should be implied by the commit message.

Can you fix the "donot" typo while you're at it?

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java [hunk 4]
```diff
@@ -476,6 +623,47 @@ public void manageChildren(
...
+    ReactShadowNode cssNodeToManage = mShadowNodeRegistry.getNode(viewTag);
+
+    ArrayList<Integer> illegalChildrendTags = new ArrayList<>();
+    for (int i = 0; i < childrenTags.size(); i++) {
+      ReactShadowNode cssNodeToAdd = mShadowNodeRegistry.getNode(childrenTags.getInt(i));
+      if (cssNodeToAdd == null) {
+        throw new IllegalViewOperationException("Trying to add unknown view tag: "
+          + childrenTags.getInt(i));
+      }
+      // If there is a RCTRawText child node, and the parent is not a ReactBaseTextShadowNode, donot add the illegal
```
Comment (hramos) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java]: This line is too long

**Comment by raphael-liu** [comment]
> Please add the copyright header to the new Java source file.

Hi, @hramos , All resolved. Thank you so much.

**Comment by raphael-liu** [comment]
Hi, @hramos . The check of test_ios was failed, but i did not modify the iOS code. And merging is blocked, could you help me with it ? Thank you.

**Comment by hramos** [comment]
@raphael-liu no worries, `test_ios` appears to be broken in trunk and unrelated to your PR.

Given this touches a ui manager, I want to make sure I get @mdvacca or @shergin to sign off on this PR, as this may affect their work on Fabric. I'll import and get it onto their queues for consideration.

File: ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java [hunk 5]
```diff
@@ -44,6 +47,14 @@
  */
 public class UIImplementation {
 
+  static {
+    try {
+      sReactBaseTextShadowNodeClass = Class.forName("com.facebook.react.views.text.ReactBaseTextShadowNode");
```
Comment (raphael-liu) [inline ReactAndroid/src/main/java/com/facebook/react/uimanager/UIImplementation.java]: Without using reflection, to break the circular dependency of buck, there will be up to 795 lines changed. And Warnings:
image

**Comment by raphael-liu** [comment]
> @raphael-liu no worries, `test_ios` appears to be broken in trunk and unrelated to your PR.
> 
> Given this touches a ui manager, I want to make sure I get @mdvacca or @shergin to sign off on this PR, as this may affect their work on Fabric. I'll import and get it onto their queues for consideration.

All right, truly appreciate your timely help.

**Comment by cpojer** [comment]
Thanks for this PR. @mdvacca reviewed it and commented "This PR increases complexity of rendering of views, this will affect performance of RN views for all RN screens. This issue was solved in React. Now React shows a proper redbox error when a RawText is added to a non-Text shadow node (on any platform)"

Therefore, I'm going to close it here.

**Comment by raphael-liu** [comment]
> Thanks for this PR. @mdvacca reviewed it and commented "This PR increases complexity of rendering of views, this will affect performance of RN views for all RN screens. This issue was solved in React. Now React shows a proper redbox error when a RawText is added to a non-Text shadow node (on any platform)"
> 
> Therefore, I'm going to close it here.

hi, @cpojer . I wonder in which version the issue was fixed in React?

**Pull request closed without merging**